include/ringmesh/basic/frame.h

```diff
@@ -147,11 +147,79 @@ namespace RINGMesh
             return origin_;
         }
 
+        vecn< DIMENSION > coords_to_local(
+            const vecn< DIMENSION >& global_coords ) const;
+
+        vecn< DIMENSION > coords_to_global(
+            const vecn< DIMENSION >& local_coords ) const
+        {
+            vecn< DIMENSION > global_coords = ( *this ).origin();
+            for( auto coord : RINGMesh::range( DIMENSION ) )
+            {
+                for( auto coor : RINGMesh::range( DIMENSION ) )
+                {
+                    global_coords[coord] +=
+                        local_coords[coor] * ( *this )[coor][coord];
+                }
+            }
+            return global_coords;
+        }
+
     private:
         vecn< DIMENSION > origin_{};
     };
     ALIAS_2D_AND_3D( ReferenceFrame );
 
+    class RINGMESH_API CartesianGridFrame3D : public ReferenceFrame< 3 >
+    {
+    public:
+        CartesianGridFrame3D() = default;
+
+        CartesianGridFrame3D( vec3 frame_origin, Frame3D frame )
+            : ReferenceFrame3D( std::move( frame_origin ), std::move( frame ) )
+        {
+            GEO::Matrix< 3, double > basechangematrix;
+            for( auto i : RINGMesh::range( 3 ) )
+            {
+                for( auto j : RINGMesh::range( 3 ) )
+                {
+                    basechangematrix( i, j ) = frame[j][i];
+                }
+            }
+            basechangematrix = basechangematrix.inverse();
+            for( auto i : RINGMesh::range( 3 ) )
+            {
+                for( auto j : RINGMesh::range( 3 ) )
+                {
+                    base_change_.origin()[i] -=
+                        basechangematrix( i, j ) * ( *this ).origin()[j];
+                    base_change_[i][j] = basechangematrix( j, i );
+                }
+            }
+        }
+
+        vec3 coords_to_local( const vec3 global_coords ) const
```

Comment:
Be carefull about your parameters. You still have a copy here. Do you see it?
This error can be found in another locations. I let you find them as a practice.